e2e/brickmux_test.go

```diff
@@ -0,0 +1,190 @@
+
+	createReq = api.VolCreateReq{
+		Name: volname2,
+		Subvols: []api.SubvolReq{
+			{
+				Type: "distribute",
+				Bricks: []api.BrickReq{
+					{PeerID: tc.gds[0].PeerID(), Path: brickPaths[3]},
+					{PeerID: tc.gds[0].PeerID(), Path: brickPaths[4]},
+				},
+			},
+		},
+		Force: true,
+	}
+	_, err = client.VolumeCreate(createReq)
+	r.Nil(err)
+
+	err = client.VolumeStart(volname2, false)
+	r.Nil(err)
+
+	// check bricks status and confirm that bricks have been multiplexed
+	// onto bricks of the first volume
+	bstatus, err = client.BricksStatus(volname2)
+	r.Nil(err)
+
+	// the pid and port variables now point to values from the old original volume
+	for _, b := range bstatus {
+		r.Equal(pid, b.Pid)
+		r.Equal(port, b.Port)
+	}
+
+	// kill the brick from first volume into which all the brick have been multiplexed
+	process, err := os.FindProcess(pid)
+	r.Nil(err, fmt.Sprintf("failed to find bricks pid: %s", err))
+	err = process.Signal(syscall.Signal(15))
+	r.Nil(err, fmt.Sprintf("failed to kill bricks: %s", err))
+
+	bstatus, err = client.BricksStatus(volname1)
+	r.Nil(err)
+	r.Equal(bstatus[0].Pid, 0)
+	r.Equal(bstatus[0].Port, 0)
+
+	// Second volume's bricks should become offline since brick from first volume has been killed
+	bstatus, err = client.BricksStatus(volname2)
+	r.Nil(err)
+	for _, b := range bstatus {
+		r.Equal(b.Online, false)
+	}
+
+	// force start the first and second volume
+	err = client.VolumeStart(volname1, true)
+	r.Nil(err)
+
+	err = client.VolumeStart(volname2, true)
+	r.Nil(err)
+
+	// first brick from first volume should now all bricks of first volume
+	// should be  be multiplexed into a new pid
+	bstatus, err = client.BricksStatus(volname1)
+	r.Nil(err)
+
+	pid = bstatus[0].Pid
+	port = bstatus[0].Port
+
+	// force start the second volume and the bricks of second volume should
+	// now be multiplexed into the pid in which bricks of first volume are multiplexed
+	bstatus, err = client.BricksStatus(volname2)
+	r.Nil(err)
+
+	for _, b := range bstatus {
+		r.Equal(pid, b.Pid)
+		r.Equal(port, b.Port)
+	}
+
+	// stop the second volume, make it incompatible for multiplexing and start it again.
+	// this should start the bricks as separate processes.
+	r.Nil(client.VolumeStop(volname2))
+
+	voloptReq := api.VolOptionReq{
+		Options: map[string]string{"write-behind.trickling-writes": "on"},
+	}
+	voloptReq.AllowAdvanced = true
+	err = client.VolumeSet(volname2, voloptReq)
+	r.Nil(err)
+
+	err = client.VolumeStart(volname2, false)
+	r.Nil(err)
+
+	bstatus, err = client.BricksStatus(volname2)
+	r.Nil(err)
+
+	// the pid and port variables point to values from the old values
+	// the bricks should have different values for pid and port as they
+	// are no longer multiplexed
+	for _, b := range bstatus {
+		r.NotEqual(pid, b.Pid)
+		r.NotEqual(port, b.Port)
+	}
+
+	r.Nil(client.VolumeStop(volname2))
+	r.Nil(client.VolumeStop(volname1))
+
+	r.Nil(client.VolumeDelete(volname2))
+	r.Nil(client.VolumeDelete(volname1))
+}
```

Comment:
I'm not seeing a test where we create ~10 odd volumes with brick mux,kill all the gluster processes and restart glusterd and check if all bricks are online with one single glusterfsd process.

Reply:
Yeah, I had some doubt in impelmenting this scenario, I was a bit skeptical whether restarting GD2 will work as expected or will it be a fresh instance with no volume records and other such things. I had a discussion with Aravinda and according to him it should work fine. So, I am working on this testcase.